File: SortVision/public/code/radix/swift/radixSort.swift
import Foundation

// Radix Sort (LSD, base 10, non-negative integers)
// Time: O(d * (n + k)), Space: O(n + k)
func radixSort(_ arr: inout [Int]) {
    guard arr.count > 1 else { return }

    guard let maxVal = arr.max() else { return }
    var exp = 1
    while maxVal / exp > 0 {
        countingSortByExp(&arr, exp)
        exp *= 10
    }
}

private func countingSortByExp(_ arr: inout [Int], _ exp: Int) {
    var output = Array(repeating: 0, count: arr.count)
    var count = Array(repeating: 0, count: 10)

    for x in arr {
        count[(x / exp) % 10] += 1
    }
    for i in 1..<10 {
        count[i] += count[i - 1]
    }

    for i in stride(from: arr.count - 1, through: 0, by: -1) {
        let digit = (arr[i] / exp) % 10
        output[count[digit] - 1] = arr[i]
        count[digit] -= 1
    }

    arr = output
}

var arr = [170, 45, 75, 90, 802, 24, 2, 66]
radixSort(&arr)
print(arr)

File: SortVision/public/code/radix/dart/radixSort.dart
class RadixSort {
  // Radix Sort (LSD, base 10, non-negative integers)
  // Time: O(d * (n + k)), Space: O(n + k)
  static void sort(List<int> arr) {
    if (arr.length <= 1) return;

    int maxVal = arr.reduce((a, b) => a > b ? a : b);
    for (int exp = 1; maxVal ~/ exp > 0; exp *= 10) {
      _countingSortByExp(arr, exp);
    }
  }

  static void _countingSortByExp(List<int> arr, int exp) {
    final output = List<int>.filled(arr.length, 0);
    final count = List<int>.filled(10, 0);

    for (final x in arr) count[(x ~/ exp) % 10]++;
    for (int i = 1; i < 10; i++) count[i] += count[i - 1];

    for (int i = arr.length - 1; i >= 0; i--) {
      final digit = (arr[i] ~/ exp) % 10;
      output[count[digit] - 1] = arr[i];
      count[digit]--;
    }

    for (int i = 0; i < arr.length; i++) arr[i] = output[i];
  }
}

void main() {
  final arr = [170, 45, 75, 90, 802, 24, 2, 66];
  RadixSort.sort(arr);
  print(arr);
}
